src/Nncase.Transform/Rules/Neutral/Fusion.cs

```diff
@@ -23,6 +30,148 @@ public abstract class FusionMaker : RewriteRule<Pattern>
+    /// <returns>TuplePattern.</returns>
+    public static Pattern MultiOutPattern(Pattern inputPattern, string outputName) => IsTuple(
+        GenerateRepeatParameters(
+            () => IsWildcardCall<EndT>(null!, null!,
+                IsWildcardCall<GetItem>(null!, null!, inputPattern))),
+        outputName);
+
+    public static Pattern EndPattern(string endCallName) => IsWildcardCall<EndT>(endCallName, null!, ComputePattern);
+
+    public static (ParameterInfo, Pattern)[] InputsPattern =>
+        typeof(DataMaker).GetField("InputsPattern").GetValue(null) as (ParameterInfo, Pattern)[];
+
+    /// <summary>
+    /// Used for construct wildcard Pattern for inputs from ParameterInfo[].
+    /// </summary>
+    /// <param name="infos">Parameter Infos.</param>
+    /// <typeparam name="BeginT" />
+    /// <returns></returns>
+    public static (ParameterInfo, Pattern)[] GenerateInputsPattern(params ParameterInfo[] infos) =>
+        infos.Select(x => (x, (Pattern)IsWildcardCall<BeginT>(null, null, (string)null))).ToArray();
+
+    /// <summary>
+    /// Get input Expr from expr of BeginT.
+    /// When you want to modify default behavior, you should override it.
+    /// </summary>
+    /// <param name="begin"></param>
+    /// <returns></returns>
+    public virtual Expr GetInputFromBegin(Expr begin)
+    {
+        return ((Call)begin).Parameters[0];
+    }
+
+    /// <summary>
+    /// Replace the old expr with new expr for each field.
+    /// </summary>
+    /// <param name="oldExpr"></param>
+    /// <param name="tupleOut"></param>
+    /// <param name="newExpr"></param>
+    /// <returns></returns>
+    public virtual IR.Tuple ReplaceTupleFields(Expr oldExpr, IR.Tuple tupleOut, Expr newExpr)
+    {
+        var newFields = tupleOut.Fields.Select(end =>
+                ReplaceFirst((Call)end, (Expr)ReplaceCallParam(
+
+                    // end is a GetItem(Call(OpT, params))
+                    // then end[0] is Call(OpT, params)
+                    (Call)((Call)end).Parameters[0],
+                    new[] { (oldExpr, newExpr) })))
+            .ToArray();
+        return tupleOut with { Fields = newFields };
+    }
+
+    protected virtual Call? GetReplace(Expr output, Call midCall, IReadOnlyList<Expr> midCallParams)
+    {
+        int idx = 0;
+
+        // get inputs for spec
+        var oldInputs = InputsPattern
+            .Select(x => x.Item1.Index)
+            .Select(i => midCallParams[i])
+            .ToArray();
+
+        // update old input and accumulate new begin and new input
+        var (newBegins, newInputs) = oldInputs
+            .Aggregate(
+                (Array.Empty<Expr>(), Array.Empty<Var>()),
+                (tuple, begin) =>
+                {
+                    var input = GetInputFromBegin(begin);
+                    var newInput = new Var($"input_{idx++}", input.CheckedType!);
+                    var newBegin = ReplaceCallParam((Call)begin, new[] { (input, (Expr)newInput) });
```

Comment:
`ReplaceCallParam` is not a good practice, because during the rewrite process, the node may be rewritten, and the leaf node of the matched node may be are before the old node, so all nodes need to from `IMatchResult`. Using `call.Parameters` for replace maybe cause rewrite error.

Reply:
impl of ReplaceCallParam is create a new call that replace old params with new param which made by ReplaceParams, it is same as 
`var beginCallParams = ReplaceParams(begin, input, newInput);`
`var newBgin = begin with { Parameters = new(beginCallParams) };`